### PR TITLE
Add configuration option `CustomTransformPatterns` to `RSpec/SpecFilePathFormat`

### PR DESCRIPTION
**Pull request by ydah**
## Motivation

For example, if you have the following files:
```ruby
# rspec_foo_spec.rb
RSpec.describe RSpecFoo do
end

# rspec_bar_spec.rb
RSpec.describe RSpecBar do
end

# rspec_baz_spec.rb
RSpec.describe RSpecBaz do
end
```

Suppose you have the following .rubocop.yml:
```yaml
RSpec/SpecFilePathFormat:
  CustomTransform:
    RSpec: rspec
```

But this would be offense:
```
❯ bundle exec rubocop rspec* --only RSpec/SpecFilePathFormat
Inspecting 3 files
CCC

Offenses:

rspec_bar_spec.rb:1:1: C: RSpec/SpecFilePathFormat: Spec path should end with r_spec_bar*_spec.rb.
RSpec.describe RSpecBar do

rspec_baz_spec.rb:1:1: C: RSpec/SpecFilePathFormat: Spec path should end with r_spec_baz*_spec.rb.
RSpec.describe RSpecBaz do

rspec_foo_spec.rb:1:1: C: RSpec/SpecFilePathFormat: Spec path should end with r_spec_foo*_spec.rb.
RSpec.describe RSpecFoo do


3 files inspected, 3 offenses detected
```

So how about something that can be specified in a pattern?

______________________________________________________________________

Before submitting the PR make sure the following are checked:

- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [x] Added tests.
- [x] Updated documentation.
- [x] Added an entry to the `CHANGELOG.md` if the new code introduces user-observable changes.
- [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

If you have modified an existing cop's configuration options:

- [x] Set `VersionChanged: "<<next>>"` in `config/default.yml`.

### CODE REVIEW

**Review by pirj — APPROVED** [review]
Wr had this option in FilePath? How comes we’ve missed copying it over to this cop whennsplitting ?
What about yhe other cop, does it need the option?
There is no way to declare this in the obsoletion conf, is there?

Thank you!

**Comment by ydah** [comment]
@pirj We didn't have this option in FilePath and had the same issue 👍🏻 

```
❯ bundle exec rubocop rspec* --only RSpec/FilePath          
Inspecting 3 files
CCC

Offenses:

rspec_bar_spec.rb:1:1: C: RSpec/FilePath: Spec path should end with r_spec_bar*_spec.rb.
RSpec.describe RSpecBar do
^^^^^^^^^^^^^^^^^^^^^^^
rspec_baz_spec.rb:1:1: C: RSpec/FilePath: Spec path should end with r_spec_baz*_spec.rb.
RSpec.describe RSpecBaz do
^^^^^^^^^^^^^^^^^^^^^^^
rspec_foo_spec.rb:1:1: C: RSpec/FilePath: Spec path should end with r_spec_foo*_spec.rb.
RSpec.describe RSpecFoo do
^^^^^^^^^^^^^^^^^^^^^^^

3 files inspected, 3 offenses detected
```

**Comment by pirj** [comment]
We probably had the same issue because ‘rspec’ was not in the defaults?
I see this option in the docs https://docs.rubocop.org/rubocop-rspec/cops_rspec.html#rspecfilepath

**Comment by ydah** [comment]
This is occurring though `RSpec: rspec` is set by default.
https://docs.rubocop.org/rubocop-rspec/cops_rspec.html#configurable-attributes-16

It also occurs when explicitly specified as follows
```
RSpec/FilePath:.
  Enabled: true
  CustomTransform: true
    RSpec: rspec
```

**Comment by pirj** [comment]
That is weird. Wondering, why?
We even have a spec for that https://github.com/rubocop/rubocop-rspec/blob/b99bf786d15d5e8b2b950307be8742980f1d09ed/spec/rubocop/cop/rspec/file_path_spec.rb#L288

Is there a guarantee that specs from this PR won’t conceal the problem ?

File: lib/rubocop/cop/rspec/spec_file_path_format.rb [hunk 1]
```diff
@@ -100,6 +110,10 @@ def expected_path(constant)
 
           File.join(
             constants.map do |name|
+              k, v = custom_transform_patterns.each.find do |k, _|
```
Comment (pirj) [inline lib/rubocop/cop/rspec/spec_file_path_format.rb:113]: ‘each’ seems redundant

Comment (pirj) [inline lib/rubocop/cop/rspec/spec_file_path_format.rb:113]: Would it make sense to give proper names like ‘pattern’ and ‘transformed’?

Comment (ydah) [reply lib/rubocop/cop/rspec/spec_file_path_format.rb]: Indeed it is, I updated this PR.

Comment (ydah) [reply lib/rubocop/cop/rspec/spec_file_path_format.rb]: Indeed it is, Thank you. Change to ‘pattern’ and ‘transformed’

**Comment by ydah** [comment]
> So this is an additional option. Wondering if we could just treat existing transforms as patterns?

I'm wondering if adding this feature as a separate additional option might be better in order to ensure backward compatibility when there are already custom settings in CustomTransform. What do you think?

File: spec/rubocop/cop/rspec/spec_file_path_format_spec.rb [hunk 2]
```diff
@@ -299,12 +299,40 @@ class Foo
...
+       'module name transformation' do
+      expect_no_global_offenses(<<-RUBY, 'rspec/some/foo/bar_spec.rb')
+        describe RSpec::Some::Foo, '#bar' do; end
+      RUBY
+    end
+  end
+
+  context 'when configured with `CustomTransformPatterns: ' \
+          '{ "RSpec" => "rspec" }`' do
+    let(:cop_config) { { 'CustomTransformPatterns' => { 'RSpec' => 'rspec' } } }
```
Comment (pirj) [inline spec/rubocop/cop/rspec/spec_file_path_format_spec.rb:323]: What i meant with reusing the option was to allow ‘“*RSpec*” => “rspec”’ config in CustomTransform.
That would still be backwards-compatible.
I’m not against the new option, but I experience some cognitive resistance due to the lack of real examples.
Would the following make good examples of partial trasgormations?
API -> api
DSL -> dsl

**Comment by ydah** [comment]
Sorry for the delay. I have updated this PR. How do you like it?

File: lib/rubocop/cop/rspec/spec_file_path_format.rb [hunk 3]
```diff
@@ -15,11 +15,21 @@ module RSpec
       #   my_class_method_spec.rb  # describe MyClass, '#method'
       #   my_class/method_spec.rb  # describe MyClass, '#method'
       #
-      # @example `CustomTransform: {RuboCop=>rubocop, RSpec=>rspec}` (default)
+      # @example `CustomTransform: {}` (default)
       #   # good
```
Comment (pirj) [inline lib/rubocop/cop/rspec/spec_file_path_format.rb:19]: Why is this good?

File: lib/rubocop/cop/rspec/spec_file_path_format.rb [hunk 4]
```diff
@@ -15,11 +15,21 @@ module RSpec
...
       #   rubocop_spec.rb          # describe RuboCop
       #   rspec_spec.rb            # describe RSpec
       #
+      # @example `CustomTransformPatterns: {RuboCop=>rubocop, RSpec=>rspec}` (default)
+      #   # bad
+      #   rspec_spec.rb            # describe RSpec
```
Comment (pirj) [inline lib/rubocop/cop/rspec/spec_file_path_format.rb:25]: Why is this bad?

File: spec/rubocop/cop/rspec/spec_file_path_format_spec.rb (hunk 2 above)
Comment (pirj) [reply spec/rubocop/cop/rspec/spec_file_path_format_spec.rb:270]: I’m still not convinced with RSpecClass.

File: lib/rubocop/cop/rspec/spec_file_path_format.rb (hunk 4 above)
Comment (ydah) [reply lib/rubocop/cop/rspec/spec_file_path_format.rb]: Ah sorry. I updated it.

File: lib/rubocop/cop/rspec/spec_file_path_format.rb (hunk 3 above)
Comment (ydah) [reply lib/rubocop/cop/rspec/spec_file_path_format.rb]: Sorry my bad. I updated it.

**Comment by ydah** [comment]
I updated this PR.

File: spec/rubocop/cop/rspec/spec_file_path_format_spec.rb [hunk 5]
```diff
@@ -233,19 +233,66 @@ class Foo
...
+      {
+        'CustomTransform' => { 'RSpec' => 'rspec' },
+        'CustomTransformPatterns' => {}
+      }
+    end
+
+    it 'registers an offense when not an exact match to custom ' \
+       'module name transformation' do
+      expect_offense(<<-RUBY, 'rspec_foo_spec.rb')
+        RSpec.describe RSpecFoo do; end
```
Comment (pirj) [inline spec/rubocop/cop/rspec/spec_file_path_format_spec.rb:247]: The best way to demo the difference between those two option would be `RSpec.describe CustomDSL` and `DSL` => `dsl` mapping. One would insist on `custom_d_s_l_*spec.rb`, the other on `custom_dsl_*spec.rb`, right?

Namespaces and `RSpecFoo` don't help understand it.

I keep losing grip about this option.

Comment (ydah) [reply spec/rubocop/cop/rspec/spec_file_path_format_spec.rb:247]: This does not appear to be an offense when followed by only one uppercase letter, as in DSL, so this is not the case that can be resolved with this option. This option may not be a common use case.

**Comment by pirj** [comment]
Please pardon my continuous inability to understand what problem this PR solves.
Is it some capitalized words with untypical conversion to snake case like DBase, RSpec, VSync, macOS that are used as **part** of multiple classes (and specs for them)? Like MacOSFileSystem MacOSKeyChain etc?
Is it widespread? How does e.g Zeitwerk deal with classes themselves?

**Comment by bquorning** [comment]
I’ll try to argue that `CustomTransform` solves this problem already. Yes, in a few cases it will require a long list, but at least that long list will be easy to understand. I find `CustomTransformPatterns` more difficult to understand.

**Comment by ydah** [comment]
That may indeed be true. Thank you. I will close this PR.

**Pull request closed without merging**